DDS-829-trash_bin

dmlond opened the pull request.

mgardnerpsu committed:
Initial draft API spec

mgardnerpsu committed:
Merge pull request #1048 from mgardnerpsu/DDS-919-modify_permissions_and_rules

Initial draft API spec

dmann committed:
File updates are no longer restricted to upload creator

dmann committed:
Renaming a file requires update_file permission

dmann committed:
Merge pull request #1055 from Duke-Translational-Bioinformatics/DDS-919-modify_permissions_and_rules

DDS-919 modify permissions and rules

dmann committed:
Merge remote-tracking branch 'upstream/DDS-923-rule_changes' into DDS-924-rule_change_update_file_permission

dmann committed:
Users can update/delete their own project permissions when another project_admin exists for the project

dmann committed:
Merge pull request #1054 from dmann/DDS-924-rule_change_update_file_permission

DDS-924 rule change update file permission

dmann committed:
Merge pull request #1056 from dmann/DDS-925-rule_change_delete_own_project_permission

DDS-925 rule change delete own project permission

dmann committed:
Merge branch 'develop' into DDS-923-rule_changes

dmlond committed:
UploadSpec expects Upload to
  - implement set_storage_container to set the storage_container to the project_id
  - call set_storage_container before_create
  - use storage_container in sub_path
Upload meets spec

ChunkSpec expects Chunk to
  - delegate storage_container to upload
  - use storage_container in sub_path

Chunk meets spec

dmlond committed:
Merge branch 'DDS-829-trash_bin' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-829-trash_bin

dmlond committed:
UploadSerializerSpect expects UploadSerializer to include storage_container
UploadSerializer meets spec

dmlond committed:
UploadSpec expects storage_container to be immutable
upload meets spec

dmlond committed:
migrates all Uploads with storage_container: nil by setting storage_container to the
project_id.

resolves #DDS-943

dmlond committed:
Containers and FileVersions implement is_purged: boolean for trashbin functionality

dmlond committed:
TrashableModel behavior defined to expect a model to validate is_purged such that
it can only be set to true on a deleted instance.

Specs for DataFile, Folder, and FileVersion expect them to
behave like a TrashableModel

TrashableModel concern implemented to add validation for is_purged

Container (Folder, DataFile), and FileVersion include TrashableModel to
meet spec

dmlond committed:
FileVersionSpec expects FileVersion not to allow is_purged to
be set to true for the current file_version of the data_file unless
the data_file itself is purged.

FileVersion meets spec

dmlond committed:
ChildPurgationJobSpec expects ChildPurgationJob to set is_purged: true for
all children of a ChildMinder
ChildPurgationJob meets spec

ChildRestorationJobSpec expects ChildRestorationJob to set is_deleted: false
for all children of a ChildMinder
ChildRestorationJob meets spec

ChildMinder behavior expects Trashable ChildMinder
  - to implement restore_children, purge_children
  - to create a ChildRestorationJobs on all pages of children
    when recently restored
  - to create a ChildPurgationJob on all pages of children
    when recently purged

FolderSpec expects Folder to behave like a Trashable ChildMinder

dmlond committed:
fixed broken cassettes

dmlond committed:
uses allow_value.for instead of setting and testing validity separately

dmann commented on spec/support/concerns/child_minder_shared_examples.rb:
```diff
@@ -1,7 +1,8 @@
   invalid_child_file_sym,
-  child_folder_sym|
+  child_folder_sym,
+  is_trashable: false|
```
The addition of this `is_trashable` parameter, and the resulting branch logic, has made the 'a ChildMinder' shared examples extremely difficult to read. I'm not sure what the best way to refactor this is, but I'm willing you pair with you to help figure it out.

dmann commented on app/models/concerns/child_minder.rb:
```diff
@@ -30,6 +58,24 @@ def delete_children(page)
+  def purge_children(page)
+    if self.class.include? TrashableModel
+      paginated_children(page).each do |child|
+        child.current_transaction = current_transaction if child.class.include? ChildMinder
```
Seems odd to test for `ChildMinder` when you're setting `#current_transaction`. Would it make sense to test for `Transactionable` instead?

dmann reviewed (changes requested):
Please refactor the ChildMinder shared examples

dmlond committed:
refactored ChildMinder shared example
  - separate Trashable ChildMinder and ChildMinder shared examples
  - expects ChildMinder to implement purge_children
  - expects purge_children to set is_deleted and is_purged to true
  - expects ChildMinder to create ChildPurgationJobs when recently deleted and not Trashable

FolderSpec expects Folder to behave_like a Trashable ChildMinder

ChildMinder meets spec

dmlond committed:
factored out shared_contexts that are common to both types of ChildMinders

dmlond committed:
deleting a project creates a ChildPurgationJob instead of a ChildDeletionJob

dmlond committed:
DataFileSpec expects DataFile to:
  - implement set_file_versions_is_deleted
  - call set_file_versions_is_deleted when is_deleted_changed
  - synchronize is_deleted state for all non-purged file_versions
DataFile meets spec

dmlond committed:
TrashableModel behavior expects is_deleted and is_purged to be
made immutable after setting is_purged to true

TrashableModel meets spec

dmlond committed:
ChunkSpec expects Chunk to implement #purge_storage which
  removes chunk object from storage_provider

dmlond committed:
UploadSpec expects Upload to implement #purge_storage
  - calls purge_storage on all chunks
  - purges object manifest from storage_provider
Upload meets spec

New Cassettes

dmlond committed:
UploadStorageRemovalJobSpec expects UploadStorageRemovalJob to
take a job_transaction and upload.id and call purge_storage on the upload
UploadStorageRemoveJob meets spec

dmlond committed:
ChildMinder modified to test is_deleted_was to make sure is_deleted is
changing from true to false

FileVersionSpec expects FileVersion to
  - behave like a job transactionable model
  - implement #manage_purge_and_restore
     - restore data_file if newly restored and data_file is deleted
     - create UploadStorageRemovalJob when newly purged
  - callback manage_purge_and_restore around update

FileVersion meets spec

dmlond committed:
Merge pull request #1057 from Duke-Translational-Bioinformatics/DDS-923-rule_changes

DDS-923 rule changes

dmann commented on spec/support/concerns/child_minder_shared_examples.rb:
```diff
@@ -35,34 +34,88 @@
+  end
+end
+
+shared_context 'common manage_children expectations' do
```
Could you change this to 'common ChildMinder#manage_children expectations'?

dmlond committed:
refactored TrashableModel into separate
  Restorable: sets up ChildMinder logic for dealing with restoration
  Purgable: sets up ChildMinder logic for dealing with purgation
added UnRestorable to set up ChildMinder logic for dealing with deletion of
an object that cannot be restored (purge children)

ChildMinder refactored to handle children with a more generic @child_job

ProjectSpec expects Project to behave like an UnRestorable ChildMinder
FolderSpec expects Folder to behave like a Purgable ChildMinder and a Restorable ChildMinder
DataFileSpec expects DataFile to behave like a Restorable and a Purgable
FileVersionSpec expects FileVersion to behave like a Restorable and a Purgable

ChildDeletionJob no longer tests the project context since only Folders delete children
ChildPurgationJob tests Folder and Project contexts since both purge their children in different
contexts

all tests pass

dmlond committed:
Merge branch 'DDS-829-trash_bin' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-829-trash_bin

dmlond committed:
added comments about implementation

dmlond committed:
Merge branch 'DDS-829-trash_bin' into DDS-947_948-FileVersionStorageRemoval

mgardnerpsu committed:
Initial draft to fix too many chunks uploaded issue.

mgardnerpsu committed:
Updated problem summary per team feedback

mgardnerpsu committed:
Added storage provider metadata to enforce chunk/file size on uploads

dmlond committed:
storage_provider now supports
  - chunk_max_number: integer
  - chunk_max_size_bytes:integer

dmlond committed:
StorageProviderSpec now expects StorageProvider to validate presence of
  - :chunk_max_number
  - :chunk_max_size_byte
StorageProvider meets spec

:storage_providers factory creates storage_provider with these values set for both default and :swift

swift.local.env sets these to standard values

dmlond committed:
the value we need to store for max_chunk_size_bytes (5368709122) is too large for int, uses a bigint instead

dmlond committed:
db:data:migrate now migrates storage_providers with nil max_chunk_number and max_chunk_size_bytes
expects ENV to be set accordingly only if there are bad storage providers

dmlond committed:
StorageProviderSerializerSpec expects StorageProviderSerializer to serialize StorageProvider with
chunk_max_number and chunk_max_size_bytes
StorageProviderSerializer meets spec

dmlond committed:
chunk size needs to be a bigint to support 5GB chunks

dmlond committed:
updated with response examples grouped with endpoint description
added deployment notes

dmlond committed:
changed size fields to generate numbers between a sensible range

dmlond committed:
UploadSpec expects Upload to
  - implement max_size_bytes (storage_provider.chunk_max_number * storage_provider.chunk_max_size_bytes)
  - validate that size is_less_than max_size_bytes
  - implement minimum_chunk_size ((size.to_f / storage_provider.chunk_max_number).ceil)

Upload meets spec

Chunks Factory creates Chunks with size between a sensible range

ChunkSpec expects Chunk to:
  - delegate chunk_max_size_bytes to storage_provider
  - delegate minimum_chunk_size to upload
  - validate size
     - presence
     - < chunk_max_size_bytes
     - >= minimum_chunk_size
     - produce custom validation message
Chunk meets spec

dmlond committed:
ChunkSpec expects Chunk to
  - implement total_chunks method which returns the current number of upload.chunks
  - validate that a chunk must not increase the total_chunks beyond the
    storage_provider.chunk_max_number
Chunk meets spec

dmlond committed:
added message to size test

dmlond committed:
Fixing merge conflicts

dmann committed:
Fixing merge conflicts

dmann committed:
Moving chunk-upload-size header test inside creatable resource block

dmann committed:
Removing ! from :payload let declaration

dmann committed:
Merge pull request #1075 from dmann/DDS-974-enforce_upload_chunk_size

DDS-974 enforce upload chunk size

dmlond committed:
Merge branch 'ua_test' of github.com:Duke-Translational-Bioinformatics/duke-data-service into try_merge_uatest_chunk

dmann committed:
Merge pull request #1076 from dmlond/try_merge_uatest_chunk

Merge ua_test chunk size enforcement to dev

dmlond committed:
This commit removes code, callbacks, and spec tests used to update the mappings of
Folders and DataFiles to support meta_property searches.

The newly minimized search no longer supports meta_property searches.

Also added tests for index_templatable_document, update_templatable_document, and
callbacks to extend our test coverage

dmann committed:
Loading and iterating over all audits for object audit summary serialization

dmann committed:
Refactor audit.user serialization into private #audit_user_info_hash method

dmann committed:
Using ActiveRecord#includes to make fewer SELECT calls

dmann committed:
Update audits do not always have a #comment["action"]

dmlond committed:
SearchableModels should callback update_index after_touch

SearchSharedExamples includes 'a SearchableModel observer'
  which expects models to enqueue an ElasticsearchIndexJob when
  they are saved

MetaPropertySpec expects MetaProperty to:
  - no longer implement index_templatable_document
    or update_templatable_document
  - no longer support callbacks to these after_save/destroy
  - behave like a SearchableModel Observer

MetaTemplateSpec expects MetaTemplate to touch the templatable it belongs_to

TagSpec expects Tag to behave like a SearchableModel observer

dmann committed:
Index db columns for faster children access

dmlond committed:
this should have gone in the last commit

dmann committed:
Merge pull request #1081 from Duke-Translational-Bioinformatics/v1_3_26-for_prod

Merging v1.3.26 to production

dmlond committed:
Merge pull request #1080 from dmlond/DDS-996_tag_elasticsearch_refresh

tag elasticsearch index management

dmann committed:
Project children endpoint takes exclude_response_fields param

dmann committed:
Folder children endpoint takes exclude_response_fields param

dmann committed:
Using ActiveRecord#includes to make fewer SELECT calls in folder children endpoint

dmann committed:
Merge branch 'DDS-1006-migration_for_table_index_creation' into DDS-1004-optimize_children_endpoint

dmann committed:
Merge branch 'DDS-1007-improve_audit_summary_serialization' into DDS-1004-optimize_children_endpoint

dmann committed:
Merge branch 'DDS-1009-custom_serialization_via_params' into DDS-1004-optimize_children_endpoint

dmann committed:
DDS-1008 #resolve do not use LIKE query when name_contains is empty

dmlond committed:
FolderSpec and DataFileSpec expect mappings to contain is_deleted: boolean
Folder and DataFile meet spec

added Implentation notes that is_deleted is indexed

dmlond committed:
FolderFilesResponseSpec expects FolderFilesResponse to filter logically deleted objects from
all queries

FolderFilesResponse meets spec

dmann committed:
Merge branch 'ua_test' into DDS-1004-optimize_children_endpoint

dmlond committed:
Merge pull request #1083 from dmann/DDS-1004-optimize_children_endpoint

DDS-1004 optimize children endpoint

dmann committed:
Merge branch 'production' into v1_3_27-for_prod

dmlond committed:
Merge pull request #1085 from Duke-Translational-Bioinformatics/v1_3_27-for_prod

Merging v1.3.27 into production

dmlond committed:
Merge pull request #1084 from dmlond/DDS-997-998_is_deleted_index

DDS-997-998-is_deleted_index

dmlond committed:
ProjectContainerElasticsearchUpdateJobSpec expects ProjectContainerElasticsearchUpdateJob to
take a transaction, project, and page and call project.update_container_elasticsearch_index_project(page)

ProjectSpec expects Project to
  - implement #manage_container_index_project: when name is changed and project has containers
    register a ProjectContainerElasticsearchUpdateJobSpec each page of containers
  - callback #manage_container_index_project after update
  - implement #update_container_elasticsearch_index_project: takes a page and performs
    a bulk elasticsearch update of the project and ancestors of each container in the page

ProjectContainerElasticsearchUpdateJob and Project meet their respective specifications

dmlond committed:
Merge branch 'develop' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-999-project_name_elasticsearch_index

dmlond committed:
cassettes needed for new Job and Job Error queue tests

dmann committed:
Removing minimum chunk size validation

dmann committed:
Merge remote-tracking branch 'upstream/ua_test' into DDS-1016-remove_min_chunk_size_validation

dmann committed:
Rewording the Response Headers section

dmann committed:
Merge pull request #1087 from dmann/DDS-1016-remove_min_chunk_size_validation

DDS-1016 remove min chunk size validation

dmlond committed:
Merge branch 'ua_test' of github.com:Duke-Translational-Bioinformatics/duke-data-service into try_merge_uatest

dmlond committed:
Merge pull request #1089 from dmlond/try_merge_uatest

Try merge uatest

dmann committed:
Merge branch 'develop' into DDS-999-project_name_elasticsearch_index

dmann committed:
Merge pull request #1086 from dmlond/DDS-999-project_name_elasticsearch_index

DDS-999-project_name_elasticsearch_index

dmlond committed:
added update_project_container_elasticsearch definition to JobsRunner
and runner to ProcFile

dmlond committed:
needs the rake worker specification

dmlond committed:
Merge pull request #1091 from dmlond/update_project_elasticsearch_jobsrunner

update_project_container_elasticsearch

dmlond committed:
creating MetaProperty mappings caused the index creation function to end in error, which
caused the rake task to end prematurely

test needed to search for file/folder by name instead of id, since we no longer index id

dmlond committed:
Merge pull request #1093 from dmlond/DDS-1025_elasticsearch_rake

DDS-1025_elasticsearch_rake

dmlond committed:
Merge branch 'develop' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-947_948-FileVersionStorageRemoval

dmlond committed:
JobsRunnerSpec now expects
  - its registered_worker_classes to include entries all ApplicationJob descendants and the MessageLogWorker
  - JobsRunner.workers_registry to include entries for all ApplicationJob descendants and the MessageLogWorker
This has the beneficial side effect that WorkersRakeSpec now fails if jobs are added without adding
entries for them to workers.rake

JobsRunner and workers.rake meet spec

dmlond committed:
purge_children, restore_children, and purge_upload job workers addedxc

dmlond committed:
fixed broken vcr cassettes

dmann commented on Procfile:
```diff
@@ -3,5 +3,8 @@ all_workers: bundle exec rake workers:all:run
 child_deletion_job: bundle exec rake workers:delete_children:run
+child_purgation_job: bundle exec rake workers:purge_children:run
+child_restoration_job: bundle exec rake workers:restore_children:run
+upload_storage_removal_job: bundle exec rake workers:purge_upload:run
```
Since you're updating the Procfile, could you add the upload completion job.

dmann reviewed (approved):
Seems fine.

dmlond committed:
WorkersRakeSpec expects ProcFile Rake Tasks to include separate heroku procs
for each Job in the JobsRunner.workers_registry

This expectation ties together with expectations that workers.rake implement
these tasks, and the expectation that new ApplicationJob.descendants are
represented in the JobsRunner.workers_registry to ensure that new jobs find
their way into workers.rake and the ProcFile.

dmlond committed:
OSX HFS provides a case insensitive file system by default, which masked this tests failure :(
